students/billy_galloway/lesson_7/assignment/hp_norton_inventory/parallel.py

```diff
@@ -0,0 +1,142 @@
+'''
+Database
+interacts with the database via files formatted from csv files
+'''
+import sys
+sys.path.append("./hp_norton_inventory")
+import logging
+import csv_handler as csvh
+from mongo_connect import *
+import threading
+from queue import Queue
+import os.path
+import pymongo
+import time
+from timeit import timeit as timer
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+def import_data(directory_name, customer_file, product_file, rental_file, db_queue):
+    ''' import data from csv files into database to be used in functions '''
+    mongo = MongoDBConnection()
+    csv_handler = csvh.CsvHandler()
+
+    ERROR_COUNT = {
+        'product': 0,
+        'customer': 0,
+        'rental': 0
+    }
+
+    INVENTORY_COUNT = {
+        'product': 0,
+        'customer': 0,
+        'rental': 0
+    }
+
+    file_names = [f'{product_file}',
+                  f'{customer_file}',
+                  f'{rental_file}']
+
+    with mongo:
+        # generate hpnorton_db
+        hpnorton_db = mongo.connection.hpnorton_db
+        collection_names = ['product', 'customer', 'rental']
+        i = 0
+
+        for name in collection_names:
+            # create collections in database
+            db_collections = hpnorton_db[name]
+            try:
+                # verify that the file exists first
+                if os.path.isfile(f'{directory_name}/{file_names[i]}'):
+                    generate_thread = threading.Thread(target=csv_handler.generate_document_list,
+                                                       args=[f'{directory_name}/{file_names[i]}', name, db_queue])
+                    # start thread and send the
+                    # return value to the queue
+                    generate_thread.start()
+                    # generate_thread.join()
```

Comment:
You do need to do a join for each of the threads you launch. Otherwise, code execution will simply continue, without waiting for running threads to finish. What you can do is group all your threads into a list:

thread_list = []
# For each thread, after they have started, do this:
thread_list.append(generate_thread)

Make sure you join all started threads before collecting your results with calls to _db_queue.get()_:

for thread in thread_list:
    thread.join()

Note that this means you don't know which thread will push its results into the queue first, so you'll need to figure out a way to mark the results so that you can identify where each set of results came from.